sql/index/pilosa/driver.go

```diff
@@ -0,0 +1,213 @@
+func (*Driver) ID() string {
+	return DriverID
+}
+
+// Create a new index.
+func (d *Driver) Create(db, table, id string, expr []sql.ExpressionHash, config map[string]string) (sql.Index, error) {
+	path, err := mkdir(d.root, db, table, id)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := index.NewConfig(db, table, id, expr, d.ID(), config)
+	err = index.WriteConfigFile(path, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqlIndex{
+		path:        path,
+		client:      d.client,
+		db:          db,
+		table:       table,
+		id:          id,
+		expressions: expr,
+	}, nil
+}
+
+// LoadAll loads all indexes for given db and table
+func (d *Driver) LoadAll(db, table string) ([]sql.Index, error) {
+	root := filepath.Join(d.root, db, table)
+
+	var indexes []sql.Index
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return filepath.SkipDir
+		}
+
+		if info.IsDir() && path != root && info.Name() != "." && info.Name() != ".." {
+			cfg, err := index.ReadConfigFile(path)
+			if err != nil {
+				return err
+			}
+
+			indexes = append(indexes, &sqlIndex{
+				path:        path,
+				client:      d.client,
+				db:          cfg.DB,
+				table:       cfg.Table,
+				id:          cfg.ID,
+				expressions: cfg.ExpressionHashes(),
+			})
+		}
+
+		return nil
+	})
+
+	return indexes, err
+}
+
+// Save the given index (mapping and bitmap)
+func (d *Driver) Save(ctx context.Context, idx sql.Index, iter sql.IndexKeyValueIter) error {
+	path, err := mkdir(d.root, idx.Database(), idx.Table(), idx.ID())
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the pilosa schema
+	schema, err := d.client.Schema()
+	if err != nil {
+		return err
+	}
+	// Create a pilosa index and frame objects in memory
+	index, err := schema.Index(indexName(idx.Database(), idx.Table(), idx.ID()))
+	if err != nil {
+		return err
+	}
+
+	frames := make([]*pilosa.Frame, len(idx.ExpressionHashes()))
+	for i, e := range idx.ExpressionHashes() {
+		frames[i], err = index.Frame(frameName(e))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Make sure the index and frames exists on the server
+	err = d.client.SyncSchema(schema)
+	if err != nil {
+		return err
+	}
+
+	mapping, err := openMapping(path)
+	if err != nil {
+		return err
+	}
+	defer mapping.close()
+
+	for colID := uint64(0); err == nil; colID++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		default:
+			var (
+				values   []interface{}
+				location []byte
+			)
+			values, location, err = iter.Next()
+			if err != nil {
+				break
+			}
+
+			for i, frm := range frames {
+				rowID, err := mapping.getRowID(frm.Name(), values[i])
+				if err != nil {
+					return err
+				}
+
+				resp, err := d.client.Query(frm.SetBit(rowID, colID))
+				if err != nil {
+					return err
+				}
+				if !resp.Success {
+					return errors.New(resp.ErrorMessage)
+				}
+			}
+			err = mapping.putLocation(index.Name(), colID, location)
+		}
+	}
```

Comment:
what happens if there is an error writing the mapping but everything is already synced with pilosa? Should we implement some kind of _transaction_ for this to ensure data integrity?

Reply:
Would be great. So far if mapping fails in Save then the function returns error.
We can cal it again and overwrite mapping and pilosa bitmap (what we can find in tests). so at the end should not be a problem just retry.